### PR TITLE
[google_maps_flutter_web] Options to disable tilt controls and configure gesture handling

### PR DESCRIPTION
**Pull request by Rexios80**
tiltControlsEnabled: On web, in satellite view, and at close zoom levels, the map displays tilt controls. This lets you disable them.

gestureHandling: On web, without greedy gesture handling the map can have an overlay saying you can only move the map with two fingers or Cmd+drag. This is not ideal, and greedy gestures lets you make that not happen. I added the other gesture handling options from the [Google Maps documentation](https://developers.google.com/maps/documentation/javascript/reference/map#MapOptions.gestureHandling) for completeness.

*List which issues are fixed by this PR. You must list at least one issue.*

[flutter/flutter/#99044](https://github.com/flutter/flutter/issues/99044)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by Rexios80** [comment]
I wasn't sure how to depend on the updated google_maps_flutter_platform_interface without depending on it form the path. Is there some fancy way to do that? Or does it need to be published first?

**Comment by Rexios80** [comment]
Oh they're supposed to be separate PRs. I guess I'll do that.

File: packages/google_maps_flutter/google_maps_flutter/CHANGELOG.md
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/CHANGELOG.md:5]: Please follow repo style; see the link in the PR checklist.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/CHANGELOG.md
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/CHANGELOG.md:3]: Same.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
@ditman Could you do the initial substantive review here, since this is primarily a web change?

**Comment by stuartmorgan** [comment]
@ditman ping on this review from triage.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:565]: Since this is web-only, and it doesn't seem to have a direct equivalent on android/iOS (they use booleans there, and the `gestureRecognizers` are more a flutter thing), let's call this:

`WebGestureHandling? webGestureHandling`

So it's very clear that it's a web-only setting that won't work elsewhere.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:87]: This should also take into account `rawOptions['rotateGesturesEnabled']` (they may be available in the web [soon?](https://developers.google.com/maps/documentation/javascript/webgl?hl=en)).

**Review by ditman — COMMENTED** [review]
This is a great addition!

I've done some comments wrt the current state of the maps JS SDK.

Also @stuartmorgan, there's a config value that might be similar across all platforms (that is named `compassEnabled` for native, which could serve as `tiltControlsEnabled` for the web.

Should we reuse the same boolean, documenting that "in web, it controls the `rotationControls` property?

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:583]: On web, you can only rotate after you toggle on 45 degree imagery (this seems to behave differently than maps.google.com, but that's probably not something we should be worrying about) which seems like a separate feature to me. At least on the projects I'm working on, we want rotation to work on android/ios, but we don't want users to be able to enable 45 degree imagery on web since that messes up our markers, polygons, etc.

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:583]: Going to have to see how the new web tilt/rotation features affect this

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: Done

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:89]: Yes, it seems that with some configuration in the GCP console, the map becomes "WebGL", and in that case, rotation and tilt become available on the web (unless disabled from the console, which I guess keeps the WebGL rendering but disables rotation and tilt, to keep things "backwards compatible") ((I'm not in the Maps team, so I don't have any info on why they decided to make this feature opt-in like they did, sorry!))

File: packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart:149]: (A fix for this just got pushed to `main`, you should get it when you rebase next. Apologies for the breakage!)

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: I'd suggest: `web45DegreeImageryEnabled`, and default to `false`. This has more to do with the 45 degree imagery feature (you need to be in satellite mode, fully zoomed in, in a supported city...) rather than the WebGL tilt feature (which looks more like the standard "tilt by dragging two fingers" in the mobile app)

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:583]: > "we don't want users to be able to enable 45 degree imagery on web" 

@Rexios80: I agree. Since AFAIK the 45deg imagery is only available on the web, we could completely disable it by always passing:

```
    tilt: 0,
    rotateControl: false,
```

to the map options when creating a web map. ([Experiment here](https://jsfiddle.net/gh/get/library/pure/googlemaps/js-samples/tree/master/dist/samples/aerial-simple/jsfiddle)).

The tilt can still be controlled by gestures or programmatically, the above two parameters seem to force to never use a tilt != 0 imagery when in satellite mode (you can get it if you go to huge zoom levels without tilt:0) and to hide the "tilt 45 degrees" button, which enables the undesirable mode.

I think that is a sane default for the plugin, that makes the web implementation more consistent with the mobile one.

(We could also add a `webEnable45DegImagery` setting that does **not** apply the default, for the people who want/need it?)

File: packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart]: I think this is dealt with

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: Will 45 degree imagery never exist on mobile? The gesture handling made sense to make very clear it's web only because that will never be on mobile, but I see no reason 45 degree imagery couldn't exist on mobile.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: @Rexios80 I don't know about Maps' plans, but I currently cannot enable 45 degree imagery in San Francisco in Google Maps itself (there's a WebGL "Globe Mode" and... a compass control on the web, haha). You need to Ctrl+Drag to change the perspective of the globe.

My feeling is that we could completely disable 45 degree imagery, and not many people would complain? (I proposed leaving the webOnlyFlag for it to be re-enabled, so we don't remove functionality from the web SDK for those who might want it)

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: Also defaulting to false isn't a good idea right now because of the behavior I've noticed with this disabled. Per my own doc comment:

> Disabling this will cause map dragging to be interrupted if the [GoogleMap] is rebuilt while dragging.

My initial guess is setting the map tilt to 0 every time the map is drawn interrupts dragging. I can try to put something together to prevent this, but I'm not sure my solution will be good enough to merge in.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: @Rexios80 I've followed internally WRT 45 degree imagery, see what we can do about it.

About the dragging being interrupted by the map re-painting, we probably should take a look. Can you prepare a minimal repro example? I guess having a map with `tilt:0` in a scrollable list is enough for the issue to happen? (I also guess that's why you added the WebGestureHandling, to try and prevent that behavior?)

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: Yeah honestly the more I play with it the more I'm not sure the plugin can deal with this itself. Setting `gmaps.MapOptions.rotateControl = false` disables the button for 45 degree imagery, but zooming in far enough still shows 45 degree imagery regardless. Setting `gmaps.MapOptions.tilt = 0` once isn't enough, because the tilt value is actually changing to 45 once zoomed in enough. And as already said, setting the tilt value interrupts map dragging, so the plugin can't just do this itself every map draw. We could maybe put something in the `onCameraMove` listener to check if the tilt value changed and force it back to 0, but that sounds vey hacky. Users of the plugin could do this themselves, but telling people to do that also seems very clunky since most everyone would want that behavior by default.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: @Rexios80 in my testing above, setting `tilt=0` and `rotateControl=false` was enough for the web sdk to not go into 45 degree mode. Then the only way to set the tilt to anything different than 0 was to call `setTilt(value)` as needed.

I've tested in this jsbin: https://jsfiddle.net/gh/get/library/pure/googlemaps/js-samples/tree/master/dist/samples/aerial-simple/jsfiddle which seems to have a functional API ready to test quickly :)

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:104]: Let me play with it more

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: Alright I think I finally got it working

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: What was holding me up is the map tilt value was reporting to be 0 by default, so I didn't think setting it to 0 when it already was would do anything 🙃

Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: Let me know what you think of the property name `fortyFiveDegreeImageryEnabled`. I really didn't want to tie it to web, because I don't see a reason 45 degree imagery couldn't exist on other platforms in the future and changing the name after merging this would be a breaking change.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: > What was holding me up is the map tilt value was reporting to be 0 by default, so I didn't think setting it to 0 when it already was would do anything 🙃

SAME, took me a few tries to add the value when the map was created.

> I don't see a reason 45 degree imagery couldn't exist on other platforms in the future

I've been told that that feature exists on "Raster" maps, but that mobile is using vector tiles (and web is going that way too). I like `fortyFiveDegreeImageryEnabled` BTW!

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart]: (This is being actively discussed in another thread, resolving this one!)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (Rexios80) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart]: @ditman Is this a breaking change? It doesn't break existing code, but it does change existing behavior.

**Comment by stuartmorgan** [comment]
What's the status here? Is this waiting for another update, or for another round of review?

**Comment by Rexios80** [comment]
This PR is waiting for another round of review

**Comment by stuartmorgan** [comment]
@ditman Ping on this re-review (from triage).

**Comment by GaryQian** [comment]
@ditman Just a friendly reminder this is ready for another round of review :)

**Comment by ditman** [comment]
I'll get this reviewed (and maybe the branch updated) on Friday, November 4th. Apologies for the delay, I got busy with a couple of other super high priority issues :(

**Comment by stuartmorgan** [comment]
@ditman Ping from triage; looks like this fell through the review cracks again.

**Comment by stuartmorgan** [comment]
Update from triage: this is waiting for @ditman to have bandwith to review maps PRs.

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!

**Pull request closed without merging**
